### PR TITLE
CPLAT-15989: Surface Event Handler Errors

### PR DESCRIPTION
**Pull request by kealjones-wk**
## Motivation
  
Native DOM behavior doesn't surface errors that occur in event handlers to the point that dispatched it. 
This can lead to false positives in tests when using `UserEvent` or `fireEvent`.

## Changes
  
- Add `eventHandlerErrorCatcher` that will listen for unhandled errors and rethrow at the site of use.
- Wrap all UserEvent and fireEvent calls to JS RTL with `eventHandlerErrorCatcher`

#### Release Notes
  
- BREAKING CHANGE: `UserEvent.*` and `fireEvent[ByName]` will now throw synchronous errors that occur in event handlers.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author:
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/react_testing_library/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/react_testing_library/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/util/event_handler_error_catcher.dart [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:html';
+
+import 'package:react/react_client/react_interop.dart' show throwErrorFromJS;
+
+/// INTENDED FOR INTERNAL USE IN REACT_TESTING_LIBRARY ONLY.
```
Comment (greglittlefield-wf) [inline lib/src/util/event_handler_error_catcher.dart:21]: There's an `@internal` annotation from the meta package we can use now to help the analyzer make sure this isn't ever exported.

File: lib/src/util/event_handler_error_catcher.dart [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+///
+/// Event Handlers triggered by a UserEvent or fireEvent may throw errors
```
Comment (greglittlefield-wf) [inline lib/src/util/event_handler_error_catcher.dart:23]: This could use a dartdoc-style opening sentence.
```suggestion
/// Calls [cb] and throws if any `window.onError` error that occur during its execution.
///
/// Event Handlers triggered by a UserEvent or fireEvent may throw errors
```

File: lib/src/util/event_handler_error_catcher.dart [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+/// This function wraps around the call to the JS RTL within UserEvent and
+/// fireEvent in order to surface the errors.
+T eventHandlerErrorCatcher<T>(T Function() cb) {
+  T result;
+  final errors = <dynamic>{};
+  final errorStream = window.onError.cast<ErrorEvent>().listen((event) {
+    errors.add(event.error);
+  });
+  result = cb();
+  errorStream.cancel();
```
Comment (greglittlefield-wf) [inline lib/src/util/event_handler_error_catcher.dart:36]: This stream subscription isn't cancelled if `cb` itself throws. Can we use a try-finally?

File: lib/src/util/event_handler_error_catcher.dart [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+/// INTENDED FOR INTERNAL USE IN REACT_TESTING_LIBRARY ONLY.
+///
+/// Event Handlers triggered by a UserEvent or fireEvent may throw errors
+/// but due to how the DOM natively works with Events, the point that dispatched
+/// the event will never surface it, leading to false positives.
+///
+/// This function wraps around the call to the JS RTL within UserEvent and
+/// fireEvent in order to surface the errors.
+T eventHandlerErrorCatcher<T>(T Function() cb) {
```
Comment (greglittlefield-wf) [inline lib/src/util/event_handler_error_catcher.dart:29]: #nit `cb` is unnecessarily abbreviated; can we use `callback` instead?

File: lib/src/util/event_handler_error_catcher.dart [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  T result;
+  final errors = <dynamic>{};
+  final errorStream = window.onError.cast<ErrorEvent>().listen((event) {
```
Comment (greglittlefield-wf) [inline lib/src/util/event_handler_error_catcher.dart:32]: #nit The name implies this is a stream, but it's a stream subscription
```suggestion
  final errorSub = window.onError.cast<ErrorEvent>().listen((event) {
```

File: lib/src/util/event_handler_error_catcher.dart [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    errors.add(event.error);
+  });
+  result = cb();
```
Comment (greglittlefield-wf) [inline lib/src/util/event_handler_error_catcher.dart:35]: I think if `cb` returns a future, we should wait until the future is completed before cancelling the subscription.

For example, if `keyboardWithDelay` throws an exception on the second keystroke and not the first. (Could we add a test for that?)

File: test/unit/util/event_handler_error.dart [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    const targetTestId = 'test-subject-11';
+    throwingHandler(event) => throw StateError('Mission Failed');
+    buildEventHandlers() => {for (final value in events) value: throwingHandler};
+    buildProps() => {
+          ...buildEventHandlers(),
+          ...additionalProps,
+          ...{defaultTestIdKey: targetTestId}
+        };
+    final targetElement = react
+        .div({}, children == null ? elementTypeToRender(buildProps()) : elementTypeToRender(buildProps(), children));
```
Comment (greglittlefield-wf) [inline test/unit/util/event_handler_error.dart:39]: #nit can we do this in a function or a setUp so it's not executed as part of the `group`?

File: test/unit/util/event_handler_error.dart [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    test('event handler error', () {
+      final view = rtl.render(targetElement as ReactElement);
+
+      expect(() => userEventTrigger(view.getByTestId(targetTestId)), throwsA(isA<StateError>()));
```
Comment (greglittlefield-wf) [inline test/unit/util/event_handler_error.dart:44]: It's possible that some unrelated code could throw a `StateError`, resulting in a false positive. Could we use an error type we know won't show up anywhere else, like `ExceptionForTesting`?

File: test/unit/user_event/clear_test.dart [hunk 9]
```diff
@@ -43,5 +45,24 @@ void main() {
...
+      final textarea = view.getByRole('textbox') as TextAreaElement;
+
+      expect(() => UserEvent.clear(textarea), throwsA(isA<UnsupportedError>()));
+    });
+
+    testEventHandlerErrors(
+      ['onChange'],
+      UserEvent.clear,
+      react.input,
+      additionalProps: {'defaultValue': 'Aaron told Sydney, Joe, Brian, and Greg that Keal was Awesome. They agreed.'},
```
Comment (greglittlefield-wf) [inline test/unit/user_event/clear_test.dart:65]: #nit
```suggestion
      additionalProps: {'defaultValue': 'Aaron told Sydney, Joe, Brian, and Greg that Keal was awesome. They agreed.'},
```

File: test/unit/user_event/clear_test.dart [hunk 10]
```diff
@@ -43,5 +45,24 @@ void main() {
...
+    test('throws when error occurs in event handler', () {
+      final view = rtl.render(react.textarea({
+        'defaultValue': 'Hello, World!',
+        'onChange': (e) {
+          throw UnsupportedError('NO CHANGES ALLOWED!');
+        }
+      }) as ReactElement);
+      final textarea = view.getByRole('textbox') as TextAreaElement;
+
+      expect(() => UserEvent.clear(textarea), throwsA(isA<UnsupportedError>()));
```
Comment (greglittlefield-wf) [inline test/unit/user_event/clear_test.dart:58]: Same comment below around using `ExceptionForTesting`.

File: test/unit/user_event/clear_test.dart [hunk 11]
```diff
@@ -43,5 +45,24 @@ void main() {
       UserEvent.clear(textarea);
       expect(textarea, hasValue(''));
     });
+
+    test('throws when error occurs in event handler', () {
```
Comment (greglittlefield-wf) [inline test/unit/user_event/clear_test.dart:49]: Also is this test redundant since we have `testEventHandlerErrors`?

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Looks good! Mainly a comment around the async methods, and some other comments here and there

File: test/unit/user_event/clear_test.dart (hunk 11 above)
Comment (kealjones-wk) [reply test/unit/user_event/clear_test.dart:49]: Yes it is, i was manually writing these tests before i realized i could just like make the shared test function thing...

File: lib/src/util/event_handler_error_catcher.dart (hunk 6 above)
Comment (KealJones) [reply lib/src/util/event_handler_error_catcher.dart]: what sucks about this is, it does work. but my test suite just like tests one error then multiple, nothing except typing had multiple possible right out the gate lol even in the typeWithDelay test i only typed one letter to not break the first test lol, ill add a typeWithDelay test that shows it working.

File: test/unit/user_event/clear_test.dart (hunk 9 above)
Comment (greglittlefield-wf) [reply test/unit/user_event/clear_test.dart]: Just correcting the casing and nothing else was the troll 😉

File: lib/src/util/event_handler_error_catcher.dart (hunk 6 above)
Comment (greglittlefield-wf) [reply lib/src/util/event_handler_error_catcher.dart]: It does? I would expect it to not work since the stream subscription would get torn down synchronously while the keyboard events are still firing asychronously

File: test/unit/util/event_handler_error.dart [hunk 12]
```diff
@@ -21,27 +21,27 @@ import 'package:react/react_client.dart' show ReactElement;
...
+      const targetTestId = 'test-subject-11';
+      throwingHandler(event) => throw ExceptionForTesting('Mission Failed');
+      buildEventHandlers() => {for (final value in events) value: throwingHandler};
+      buildProps() => {
+            ...buildEventHandlers(),
+            ...additionalProps,
+            ...{defaultTestIdKey: targetTestId}
+          };
+      final targetElement = react
+          .div({}, children == null ? elementTypeToRender(buildProps()) : elementTypeToRender(buildProps(), children));
```
Comment (greglittlefield-wf) [inline test/unit/util/event_handler_error.dart]: This isn't quite what I meant 😄.

Instead of being executed in the nested group, now it's being executed in its parent group. Could we move it to a function or a `setUp` so it's only run when the tests are run, as opposed to when the tests are being declared?

Comment (KealJones) [reply test/unit/util/event_handler_error.dart]: Im not sure i understand. Why does it matter if it is when the test is declared vs when its run? It's a final variable, it's the same thing in both tests, it doesn't need to be run before every test as setUp would do, in fact it would then not be final anymore and would be rebuilding the same thing before every test. Why is that better?

Comment (greglittlefield-wf) [reply test/unit/util/event_handler_error.dart]: Sorry, I wasn't doing a good job of explaining why I made this suggestion.

In general, we should avoid doing computations outside of test blocks because:
1. The code in `group` blocks gets run at startup every time you run tests, even if you're running a subset of the tests (by test name filtering via `-n`/`-N`, tag filtering, etc.). The `group`s themselves aren't skipped and all of them get run, it's just the `test` blocks and corresponding setUp/etc. blocks that get skipped. So, anything that's done in `group`s adds overhead to test startup, and also can get run unnecessarily when tests are being filtered.

    This one instance of creating a ReactElement isn't a big deal, but if we get in the habit of doing it in more places and with more expensive computations, it can add up and slow things down.

1. Related, if something goes wrong when initializing that variable, it's nicer if it gets run as part of the test it's being used in, so that the error message can be more easily associated with the test.


So I think either making targetElement a function instead of a variable would probably be best here.

File: test/unit/user_event/type_test.dart [hunk 13]
```diff
@@ -197,12 +199,32 @@ void _typeTestHelper({bool hasDelay = false, bool isTextArea = false}) {
...
+
+  group('', () {
     testEventHandlerErrors(
       ['onInput'],
       (el) => hasDelay ? UserEvent.typeWithDelay(el, 'K', Duration(hours: 500)) : UserEvent.type(el, 'K'),
       isTextArea ? react.textarea : react.input,
     );
-  });
+
+    test('multiple event handler errors on every letter', () {
```
Comment (greglittlefield-wf) [inline test/unit/user_event/type_test.dart:211]: I think a better description would be "event handlers that occur on delayed keystrokes" or something, since that's really what it's testing

File: test/unit/user_event/type_test.dart [hunk 14]
```diff
@@ -197,12 +199,32 @@ void _typeTestHelper({bool hasDelay = false, bool isTextArea = false}) {
...
+      const stringToTest = 'Hello There';
+      final view = rtl.render((isTextArea ? react.textarea : react.input)({
+        defaultTestIdKey: 'event-handle-error-tester',
+        'onKeyUp': (e) => throw ExceptionForTesting('ow'),
```
Comment (greglittlefield-wf) [inline test/unit/user_event/type_test.dart:215]: Can we add some comment here saying that this is `keyUp` since it happens after the delay, and won't throw synchronously?

File: lib/src/dom/fire_event.dart [hunk 15]
```diff
@@ -69,8 +70,10 @@ import '../user_event/user_event.dart';
 /// See: <https://testing-library.com/docs/dom-testing-library/api-events/#fireevent>
 ///
 /// {@category UserActions}
+bool fireEvent(Element element, Event event) => eventHandlerErrorCatcher(() => _fireEvent(element, event));
```
Comment (greglittlefield-wf) [inline lib/src/dom/fire_event.dart:73]: Oh I just thought... could we add a small note in each of these doc comments saying that, unlike the JS implementation, any uncaught errors thrown during event propagation will get rethrown at the end?

I think it's important to document this behavior since it's different than in the JS

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
A couple more really small things otherwise LGTM

File: test/unit/util/event_handler_error.dart (hunk 12 above)
Comment (kealjones-wk) [reply test/unit/util/event_handler_error.dart]: Oohhh KK that makes so much sense. Thank you for clarifying! I never knew that this was the case and for sure do not want that to become an issue. I will remember this lesson going forward :)

File: test/unit/user_event/type_test.dart (hunk 14 above)
Comment (greglittlefield-wf) [reply test/unit/user_event/type_test.dart:215]: Sorry, I didn't read this test properly 😓 ... disregard this comment, sorry about that!

**Review by greglittlefield-wf — APPROVED** [review]
+10

**Review by aaronlademann-wf — APPROVED** [review]
+1 / Semver +1

**Comment by aaronlademann-wf** [comment]
@Workiva/release-management-pp

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM
